### PR TITLE
Use mo_pack for WGDOS unpacking/packing

### PR DESCRIPTION
**Pull request by lbdreyer**

### CODE REVIEW

File: lib/iris/experimental/um.py [hunk 1]
```diff
@@ -405,10 +405,16 @@ def read_data(self):
...
                 data = data.reshape(rows, cols)
             elif lbpack == 1:
-                from iris.fileformats.pp_packing import wgdos_unpack
+                try:
+                    from mo_pack import unpack_wgdos
+                except ImportError:
+                    msg = 'iris.fileformats.pp_packing.wgdos_unpack has been ' \
+                          'deprecated. Unpacking PP fields with LBPACK of {} ' \
+                          'now requires mo_pack to be installed'.format(lbpack)
+                    raise NotImplementedError(msg)
```
Comment (rhattersley) [inline lib/iris/experimental/um.py]: For this to be a deprecation the old functionality should still be available.

File: lib/iris/fileformats/pp.py [hunk 2]
```diff
@@ -888,8 +893,14 @@ def _data_bytes_to_shaped_array(data_bytes, lbpack, boundary_packing,
...
-        data = pp_packing.wgdos_unpack(data_bytes, data_shape[0],
-                                       data_shape[1], mdi)
+        if mo_pack is not None:
+            data = mo_pack.unpack_wgdos(data_bytes,
+                                        data_shape[0], data_shape[1], mdi)
+        else:
+            msg = 'iris.fileformats.pp_packing.wgdos_unpack has been ' \
+                  'deprecated. Unpacking PP fields with LBPACK of {} ' \
+                  'now requires mo_pack to be installed'.format(lbpack.n1)
+            raise NotImplementedError(msg)
```
Comment (rhattersley) [inline lib/iris/fileformats/pp.py]: For this to be a deprecation the old functionality should still be available.

File: lib/iris/experimental/um.py (hunk 1 above)
Comment (lbdreyer) [inline lib/iris/experimental/um.py]: We have chosen not to use deprecations here since it is "experimental"

Comment (rhattersley) [inline lib/iris/experimental/um.py]: :+1:

File: lib/iris/experimental/um.py [hunk 3]
```diff
@@ -405,11 +409,16 @@ def read_data(self):
...
-                data_bytes = self.source.read(data_size)
-                data = wgdos_unpack(data_bytes, field.lbrow, field.lbnpt,
-                                    field.bmdi)
+                if mo_pack is not None:
+                    data_size = ((field.lbnrec * 2) - 1) * _WGDOS_SIZE
+                    data_bytes = self.source.read(data_size)
+                    data = mo_pack.unpack_wgdos(data_bytes, field.lbrow,
+                                                field.lbnpt, field.bmdi)
+                else:
+                    msg = 'iris.fileformats.pp_packing.wgdos_unpack has been ' \
```
Comment (pelson) [inline lib/iris/experimental/um.py]: I don't think this is representative of the exception. ``mo_pack is needed to read packed data`` would be a suitable message.

File: lib/iris/fileformats/pp.py [hunk 4]
```diff
@@ -888,10 +897,24 @@ def _data_bytes_to_shaped_array(data_bytes, lbpack, boundary_packing,
     if lbpack.n1 in (0, 2):
         data = np.frombuffer(data_bytes, dtype=data_type)
     elif lbpack.n1 == 1:
-        data = pp_packing.wgdos_unpack(data_bytes, data_shape[0],
-                                       data_shape[1], mdi)
+        if mo_pack is not None:
+            data = mo_pack.unpack_wgdos(data_bytes,
+                                        data_shape[0], data_shape[1], mdi)
+        elif pp_packing is not None:
+            msg = 'The WGDOS unpacking functionality is deprecated and will ' \
```
Comment (pelson) [inline lib/iris/fileformats/pp.py]: Perhaps include the name in here. ```iris.fileformats.pp_packing``` has been deprecated and will ...

File: lib/iris/fileformats/pp.py [hunk 5]
```diff
@@ -42,7 +42,16 @@
 import iris.fileformats.pp_rules
 import iris.coord_systems
 import iris.proxy
```
Comment (pelson) [inline lib/iris/fileformats/pp.py]: This import may now be redundant.

File: lib/iris/fileformats/pp.py [hunk 6]
```diff
@@ -1403,9 +1438,11 @@ def save(self, file_handle):
         # the data itself
         if lb[self.HEADER_DICT['lbpack'][0]] == 0:
             data.tofile(pp_file)
+        elif lb[self.HEADER_DICT['lbpack'][0]] == 1:
+            pp_file.write(packed_data)
         else:
             msg = 'Writing packed pp data with lbpack of {} ' \
-                'is not supported.'.format(lb[self.HEADER_DICT['lbpack'][0]])
+                  'is not supported.'.format(lb[self.HEADER_DICT['lbpack'][0]])
             raise NotImplementedError(msg)
```
Comment (pelson) [inline lib/iris/fileformats/pp.py]: I'm not sure about this exception. I'd look at implementing either a ValueError (generic), or an ImportError (as it means we don't have the necessary package).

**Comment by pelson** [comment]
:+1: other than those comments.

File: lib/iris/fileformats/pp.py (hunk 6 above)
Comment (lbdreyer) [inline lib/iris/fileformats/pp.py]: To me, ImportError would suggest that a package that packed data with lbpacks that aren't 0 or 1 exists. As far as I'm aware, this doesn't currently exist. 
It appears that you should only get ImportErrors  when you've tried to import a package (https://docs.python.org/2/library/exceptions.html#exceptions.ImportError)

NotImplementedError was what we had before when we didn't even have wgdos packing, so surely we have the same situation here?

Comment (lbdreyer) [inline lib/iris/fileformats/pp.py]: Perhaps: http://scitools.org.uk/iris/docs/dev/iris/iris/exceptions.html#iris.exceptions.NotYetImplementedError?

File: lib/iris/fileformats/pp.py [hunk 7]
```diff
@@ -1403,10 +1437,12 @@ def save(self, file_handle):
...
         if lb[self.HEADER_DICT['lbpack'][0]] == 0:
             data.tofile(pp_file)
+        elif lb[self.HEADER_DICT['lbpack'][0]] == 1:
+            pp_file.write(packed_data)
         else:
             msg = 'Writing packed pp data with lbpack of {} ' \
-                'is not supported.'.format(lb[self.HEADER_DICT['lbpack'][0]])
-            raise NotImplementedError(msg)
+                  'is not supported.'.format(lb[self.HEADER_DICT['lbpack'][0]])
+            raise ValueError(msg)
```
Comment (pelson) [inline lib/iris/fileformats/pp.py]: My bad. This needs reverting.

**Comment by QuLogic** [comment]
I'm guessing there's been no release of `mo_pack` yet, though?

**Comment by pelson** [comment]
> I'm guessing there's been no release of mo_pack yet, though?

No, not yet. Your PR helps bring it forwards though, so thanks @QuLogic. There is no real reason for us not to cut a v0.1 in very short order.

**Comment by pelson** [comment]
Really pleasing @lbdreyer. Great work - we can now write packed PP files! :tada:

**Pull request merged**